TestFoundation/TestOperationQueue.swift

```diff
@@ -168,34 +168,42 @@ class TestOperationQueue : XCTestCase {
 class AsyncOperation: Operation {
 
     private let queue = DispatchQueue(label: "async.operation.queue")
-    private let lock = NSLock()
+    private let lock = NSRecursiveLock()
 
     private var _executing = false
     private var _finished = false
 
     override internal(set) var isExecuting: Bool {
         get {
+            self.lock.lock()
+            defer { self.lock.unlock() }
```

Comment:
using the defer here does miss out on the thread safety consideration that consumers should think about:
```
var isExecuting: Bool {
    get {
        lock.lock()
        let wasExecuting = _executing
        lock.unlock()
        return wasExecuting
    }
}
```

it is a bit cheeky but it gets the point across that the return value should only be a perspective of the potential past.